File: lib/aerox/widgets/Sizer.cpp
#include "aerox/widgets/Sizer.hpp"

namespace aerox::widgets {
void Sizer::SetWidth(const std::optional<float> &width) {
  _width = width;
  InvalidateCachedSize();
}

void Sizer::SetHeight(const std::optional<float> &height) {
  _height = height;
  InvalidateCachedSize();
}

std::optional<float> Sizer::GetWidth() const {
  return _width;
}

std::optional<float> Sizer::GetHeight() const {
  return _height;
}

std::optional<uint32_t> Sizer::GetMaxSlots() const {
  return 1;
}

void Sizer::Draw(
    WidgetFrameData *frameData, const DrawInfo info) {

  if(!GetDrawRect().HasIntersection(info.clip)) {
    return;
  }

  const auto clip = info.clip.Clone().Clamp(GetDrawRect());
  
  if(!clip.HasSpace()) {
    return;
  }
  
  if(const auto slot = _slots.try_index(0); slot.has_value()) {
    if(const auto widget = slot.value()->GetWidget().lock()) {
      auto drawRect = widget->UpdateDrawRect(Rect().Offset(GetDrawRect().GetPoint()).SetSize(GetDrawRect().GetSize()));
      widget->Draw(frameData,{this,clip});
    }
  }
}

Size2D Sizer::ComputeDesiredSize() const {
  if(const auto slot = _slots.try_index(0); slot.has_value()) {
    auto childSize = slot.value()->GetWidget().lock()->GetDesiredSize();
    return {_width.value_or(childSize.width),_height.value_or(childSize.height)};
  }
  
  
  return {};
}

}

File: aerox.Runtime.Widgets/Defaults/Containers/Sizer.cs
using aerox.Runtime.Math;

namespace aerox.Runtime.Widgets.Defaults.Containers;

public class Sizer : Container
{
    private float? _heightOverride;
    private float? _widthOverride;

    public Sizer(Widget child) : base(child)
    {
    }

    public Sizer()
    {
    }

    public float? WidthOverride
    {
        get => _widthOverride;
        set
        {
            _widthOverride = value;
            CheckSize();
        }
    }

    public float? HeightOverride
    {
        get => _heightOverride;
        set
        {
            _heightOverride = value;
            CheckSize();
        }
    }

    public override Size2d ComputeDesiredSize()
    {
        if (slots.Count == 0) return new Size2d();

        var slot = slots[0];
        var desiredSize = slot.GetWidget().GetDesiredSize();
        return new Size2d(WidthOverride ?? desiredSize.Width, HeightOverride ?? desiredSize.Height);
    }


    public override void Draw(WidgetFrame frame, DrawInfo info)
    {
        foreach (var slot in slots.ToArray()) slot.GetWidget().Draw(frame, info.AccountFor(slot.GetWidget()));
    }

    public override uint GetMaxSlots()
    {
        return 1;
    }

    protected override void ArrangeSlots(Size2d drawSize)
    {
        if (slots.Count == 0) return;

        var slot = slots[0];
        slot.GetWidget().SetRelativeOffset(new Vector2<float>(0, 0));
        slot.GetWidget().SetDrawSize(drawSize.Clone());
    }
}